File: sudoku.c
#include <stdio.h>

int board[11][11];
int qx[3] = {0, 1, -1};
int qy[3] = {1, 0, 0};

void print_board();
void solve(int x, int y);
int solved();
int check_square(int num, int x, int y);
int check_lines(int num, int x, int y);

int main()
{
    for (int i = 0; i < 11; i++)
    {
        for (int j = 0; j < 11; j++)
        {
            board[i][j] = -1;
        }
        
    }
    
    FILE *b = fopen("sudoku_board.txt", "r");
    for (int i = 1; i < 10; i++)
    {
        for (int j = 1; j < 10; j++)
        {
            fscanf(b, "%d", &board[i][j]);
        }
    }
    for (int i = 0; i < 11; i++)
    {
        for (int j = 0; j < 11; j++)
        {
            printf("%2d ", board[i][j]);
        }
        printf("\n");
        
    }
    solve(1,1);
    
}

void print_board()
{
    for (int i = 1; i < 10; i++)
    {
        for (int j = 1; j < 10; j++)
        {
            printf("%d ", board[i][j]);
            if(j % 3 == 0 && j != 9)
                printf(" ");
        }
        if(i % 3 == 0)
            printf("\n");
        printf("\n");
    }
}

int check_lines(int n, int x, int y)
{
    for (int i = 1; i < 10; i++)
        if (board[x][i] == n)
            return 0;

    for (int i = 1; i < 10; i++)
        if (board[i][y] == n)
            return 0;

    return 1;    
}

int check_square(int n, int x, int y)
{
    int xp = (int)((x-1) / 3) * 3 + 1;
    int yp = (int)((y-1) / 3) * 3 + 1;
    for (int i = xp; i < xp + 3; i++)
        for (int j = yp; j < yp + 3; j++)
            if (board[i][j] == n)
                return 0;

    return 1;
    
}

int solved()
{
    for (int i = 1; i < 10; i++)
        for (int j = 1; j < 10; j++)
            if(board[i][j] == 0)
                return 0;

    return 1;
}

void solve(int x, int y)
{
    if(board[x][y] != -1){
        if(board[x][y] == 0){
            for (int num = 1; num < 10; num++)
            {
                if(check_lines(num, x, y) && check_square(num, x, y))
                {
                    board[x][y] = num;
                    print_board();
                    for (int i = 0; i < 3; i++)
                    {
                        solve(x + qx[i], y + qy[i]);
                    }
                    board[x][y] == 0;
                    if(solved())
                    {
                        print_board();
                        exit(1);
                    }
                }
            }
        }
        else
        {
            for (int i = 0; i < 2; i++)
            {
                solve(x + qx[i], y + qy[i]);
            }
        }
        
    }
}

File: sudoku.py
import pygame
import time

with open("sudoku_board.txt") as sFile:
    board = [line.split() for line in sFile]

board = [[int(i) for i in line] for line in board]

zeros = []
for i in range(9):
    for j in range(9):
        if(board[i][j] == 0):
            zeros.append((i, j))

def check(n, pair):
    i, j = pair

    if n in board[i]:
        return 0

    for line in board:
        if line[j] == n:
            return 0

    initial_i, initial_j = i // 3 * 3, j // 3 * 3
    
    for x in range(initial_i, initial_i+3):
        for y in range(initial_j, initial_j+3):
            if(board[x][y] == n):
                return 0                
    return 1

def solved():
    for line in board:
        for i in line:
            if i == 0:
                return 0
    return 1

def print_board():
    for i in range(9):
        for j in range(9):
            print(board[i][j], end = ' ')

            if ((j+1) % 3 == 0) and (j != 9):
                print(' ', end='')

        if((i+1) % 3 == 0):
            print('')
        print('')

def solve(i):
    for num in range(1, 10):
        x = zeros[i][0]
        y = zeros[i][1]

        if check(num, zeros[i]):

            pygame.draw.rect(gameDisplay, (0, 255, 0), ((y+0.1)*block_size, (x+0.1)*block_size, block_size-10, block_size-10))
            textsurface = myfont.render(str(num), False, (0, 0, 0))
            gameDisplay.blit(textsurface, ((y+0.36)*(block_size), (x+0.2)*(block_size)))
            pygame.display.update()
            time.sleep(0.1)
            
            board[zeros[i][0]][zeros[i][1]] = num

            if i < (len(zeros) - 1):
                pygame.draw.rect(gameDisplay, (255, 255, 255), ((y+0.1)*block_size, (x+0.1)*block_size, block_size-10, block_size-10))
                textsurface = myfont.render(str(num), False, (0, 0, 0))
                gameDisplay.blit(textsurface, ((y+0.36)*(block_size), (x+0.2)*(block_size)))
                solve(i+1)
            else:
                print_board()
                pygame.draw.rect(gameDisplay, (255, 255, 255), ((y+0.1)*block_size, (x+0.1)*block_size, block_size-10, block_size-10))
                textsurface = myfont.render(str(num), False, (0, 0, 0))
                gameDisplay.blit(textsurface, ((y+0.36)*(block_size), (x+0.2)*(block_size)))
                time.sleep(100)
            board[zeros[i][0]][zeros[i][1]] = 0
            pygame.draw.rect(gameDisplay, (255, 255, 255), ((y+0.1)*block_size, (x+0.1)*block_size, block_size-10, block_size-10))
        pygame.draw.rect(gameDisplay, (255, 255, 255), ((y+0.1)*block_size, (x+0.1)*block_size, block_size-10, block_size-10))
        

# def main():
pygame.init()

gameDisplay = pygame.display.set_mode((459,459))
pygame.display.set_caption('Sudoku')

background = pygame.Surface(gameDisplay.get_size())
background.fill((255,255,255))
background = background.convert()
gameDisplay.blit(background, (0, 0))

block_size = gameDisplay.get_size()[0] // 9

for x in range(9):
    for y in range(9):
        pygame.draw.rect(gameDisplay, (0, 0, 0), (x*block_size, y*block_size, block_size, block_size), 1)

myfont = pygame.font.SysFont('Arial', 25)
print(board)
for x in range(9):
    for y in range(9):
        if board[x][y] != 0:
            textsurface = myfont.render(str(board[x][y]), False, (0, 0, 0))
            gameDisplay.blit(textsurface, ((y+0.36)*(block_size), (x+0.2)*(block_size)))
for i in range(1,3):
    pygame.draw.line(gameDisplay, (0, 0, 0), (3*i*block_size, 0), (3*i*block_size, 9*i*block_size), 4)
for i in range(1,3):
    pygame.draw.line(gameDisplay, (0, 0, 0), (0, 3*i*block_size), (9*i*block_size, 3*i*block_size), 4)
clock = pygame.time.Clock()

running = True

while running:

    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            running = False

    solve(0)
    
    pygame.display.update()
    clock.tick(60)
    

pygame.quit()
quit()

# if __name__ == "__main__":
#     main()